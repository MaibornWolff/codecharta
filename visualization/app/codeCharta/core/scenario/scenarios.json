[
  {
    "name": "Complexity",
    "settings": {
      "neutralColorRange": {
        "from": 20,
        "to": 40,
        "flipped": false
      },
      "areaMetric": "rloc",
      "heightMetric": "mcc",
      "colorMetric": "mcc",
      "mode": "KindOfMap.Single",
      "amountOfTopLabels": 1,
      "scaling": {
        "x": 1,
        "y": 1,
        "z": 1
      },
      "camera": {
        "x": 0,
        "y": 300,
        "z": 1000
      },
      "margin": 1,
      "deltaColorFlipped": false,
      "enableEdgeArrows": true,
      "hideFlatBuildings": true,
      "maximizeDetailPanel": false,
      "invertHeight": false,
      "dynamicMargin": true,
      "isWhiteBackground": false,
      "blacklist":[]
    },
    "autoFitCamera": true
  },
  {
    "name": "Average Complexity*",
    "settings": {
      "neutralColorRange": {
        "from": 20,
        "to": 40,
        "flipped": false
      },
      "areaMetric": "unary",
      "heightMetric": "Average Complexity*",
      "colorMetric": "Average Complexity*",
      "mode": "KindOfMap.Single",
      "amountOfTopLabels": 1,
      "scaling": {
        "x": 1,
        "y": 1,
        "z": 1
      },
      "camera": {
        "x": 0,
        "y": 300,
        "z": 1000
      },
      "margin": 1,
      "deltaColorFlipped": false,
      "enableEdgeArrows": true,
      "hideFlatBuildings": true,
      "maximizeDetailPanel": false,
      "invertHeight": false,
      "dynamicMargin": true,
      "isWhiteBackground": false,
      "blacklist":[]
    },
    "autoFitCamera": true
  },
  {
    "name": "Coverage",
    "settings": {
      "neutralColorRange": {
        "from": 50,
        "to": 75,
        "flipped": true
      },
      "areaMetric": "rloc",
      "heightMetric": "mcc",
      "colorMetric": "line_coverage",
      "mode": "KindOfMap.Single",
      "amountOfTopLabels": 1,
      "scaling": {
        "x": 1,
        "y": 1,
        "z": 1
      },
      "camera": {
        "x": 0,
        "y": 300,
        "z": 1000
      },
      "margin": 1,
      "deltaColorFlipped": false,
      "enableEdgeArrows": true,
      "hideFlatBuildings": true,
      "maximizeDetailPanel": false,
      "invertHeight": false,
      "dynamicMargin": true,
      "isWhiteBackground": false,
      "blacklist":[]
    },
    "autoFitCamera": true
  },
  {
    "name": "Code Churn",
    "settings": {
      "neutralColorRange": {
        "from": 10,
        "to": 30,
        "flipped": false
      },
      "areaMetric": "rloc",
      "heightMetric": "abs_code_churn",
      "colorMetric": "weeks_with_commits",
      "mode": "KindOfMap.Single",
      "amountOfTopLabels": 1,
      "scaling": {
        "x": 1,
        "y": 1,
        "z": 1
      },
      "camera": {
        "x": 0,
        "y": 300,
        "z": 1000
      },
      "margin": 1,
      "deltaColorFlipped": false,
      "enableEdgeArrows": true,
      "hideFlatBuildings": true,
      "maximizeDetailPanel": false,
      "invertHeight": false,
      "dynamicMargin": true,
      "isWhiteBackground": false,
      "blacklist":[]
    },
    "autoFitCamera": true
  }
]