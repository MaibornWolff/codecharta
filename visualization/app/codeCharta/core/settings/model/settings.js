/**
 * This class defines the available set of settings in the CodeCharta application
 */
export class Settings {

    /**
     * @constructor
     * @param {CodeMap} map
     * @param {Range} neutralColorRange
     * @param {string} areaMetric
     * @param {string} heightMetric
     * @param {string} colorMetric
     * @param {boolean} deltas
     * @param {Number} amountOfTopLabels
     * @param {Scale} scaling
     * @param {Scale} camera
     * @param {Number} margin
     */
    constructor(map, neutralColorRange, areaMetric, heightMetric, colorMetric, deltas, amountOfTopLabels, scaling, camera, margin) {

        /**
         * currently selected map
         * @type {CodeMap}
         */
        this.map = map;

        /**
         * currently set neutral color range. This defines the yellow color range and the whether higher or lower values are green or red.
         * @type {Range}
         */
        this.neutralColorRange = neutralColorRange;

        /**
         * currently selected area metric
         * @type {string}
         */
        this.areaMetric = areaMetric;

        /**
         * currently selected heightm metric
         * @type {string}
         */
        this.heightMetric = heightMetric;

        /**
         * currently selected color metric
         * @type {string}
         */
        this.colorMetric = colorMetric;

        /**
         * delta flag
         * @type {boolean}
         */
        this.deltas = deltas;

        /**
         * number of highest buildings with labels
         * @type {Number}
         */
        this.amountOfTopLabels = amountOfTopLabels;

        /**
         * scaling settings
         * @type {Scale}
         */
        this.scaling = scaling;

        /**
         * camera position
         * @type {Scale}
         */
        this.camera = camera;

        /**
         * margin between buildings
         * @type {number}
         */
        this.margin = margin;

    }

    /**
     * Imports the given settings values without replacing this object with the given one.
     * @param {Settings} settings given settings
     */
    importSettingValues(settings){
        this.map = settings.map;
        this.neutralColorRange = settings.neutralColorRange;
        this.areaMetric = settings.areaMetric;
        this.heightMetric = settings.heightMetric;
        this.colorMetric = settings.colorMetric;
        this.deltas = settings.deltas;
        this.amountOfTopLabels = settings.amountOfTopLabels;
        this.scaling = settings.scaling;
        this.camera = settings.camera;
        this.margin = settings.margin;
    }

}
