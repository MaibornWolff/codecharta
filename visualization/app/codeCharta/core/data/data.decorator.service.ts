"use strict";

import * as d3 from "d3";
import {CodeMap, CodeMapNode} from "./model/CodeMap";
import {HierarchyNode} from "d3-hierarchy";

/**
 * Decorates the data structure with artificial metrics
 */
export class DataDecoratorService {

    /* @ngInject */
    constructor() {

    }

    /**
     * Decorates the map with the unary metric. This metric is always 1 to allow the same area on all buildings.
     * @param {CodeMap} map
     */
    public decorateMapWithUnaryMetric(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            let descendants: HierarchyNode<CodeMapNode>[] = root.descendants();

            for (let j = 0; j < descendants.length; j++) {
                if (!descendants[j].data.attributes) {
                    descendants[j].data.attributes = {};
                }
                Object.assign(descendants[j].data.attributes, {unary: 1});
            }

        }
    }

    public decorateMapWithVisibleAttribute(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                node.data.visible = true;
            });

        }
    }

    public decorateMapWithOriginAttribute(map: CodeMap) {

        if (map && map.root) {

            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                node.data.origin = map.fileName;
            });

        }
    }

    public decorateMapWithPathAttribute(map: CodeMap) {
        if (map && map.root) {
            let root = d3.hierarchy<CodeMapNode>(map.root);
            root.each((node) => {
                let path = root.path(node);
                let pathString = "";
                path.forEach((pnode) => {
                    pathString += "/" + pnode.data.name;
                });
                node.data.path = pathString;
            });
        }
    }

    public decorateLeavesWithMissingMetrics(maps: CodeMap[], metrics: string[]) {

        maps.forEach((map) => {
            if (map && map.root) {

                let root = d3.hierarchy<CodeMapNode>(map.root);
                root.leaves().forEach((node) => {

                    //make sure attributes exist
                    this.createAttributesIfNecessary(node);

                    //create Metrics
                    for (let i = 0; i < metrics.length; i++) {
                        let metric = metrics[i];
                        if (!node.data.attributes.hasOwnProperty(metric)) {
                            node.data.attributes[metric] = 0;
                        }
                    }

                });

            }
        });

    }

    public decorateParentNodesWithMeanAttributesOfChildren(maps: CodeMap[], metrics: string[]) {

        maps.forEach((map) => {
            if (map && map.root) {

                let root = d3.hierarchy<CodeMapNode>(map.root);
                root.each((node) => {
                    this.decorateNodeWithChildrenMeanMetrics(node, metrics);

                });

            }
        });
    }

    public decorateNodeWithChildrenMeanMetrics(node, metrics: string[]) {
        //make sure attributes exist
        this.createAttributesIfNecessary(node);

        //if attributes is empty define property for each possible metric as a mean function of child metrics
        for (let i = 0; i < metrics.length; i++) {
            let metric = metrics[i];
            if (!node.data.attributes.hasOwnProperty(metric) && node.data.children && node.data.children.length > 0) {
                this.defineAttributeAsMeanMethod(node, metric);
            }
        }

    }

    private defineAttributeAsMeanMethod(node, metric: string) {
        Object.defineProperty(node.data.attributes, metric, {
            enumerable: true,
            get: function () {
                let count = 0;
                let sum = 0;
                let l = node.leaves();
                for (; count < l.length; count++) {
                    sum += l[count].data.attributes[metric];
                }
                return sum / count;
            }
        });
    }

    private createAttributesIfNecessary(node) {
        if (!node.data.attributes) {
            node.data.attributes = {};
        }
    }
}
