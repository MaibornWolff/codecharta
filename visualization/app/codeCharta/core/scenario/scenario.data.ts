import {Scenario} from "./scenario.service";
import {CodeMap} from "../data/model/CodeMap";
import {STATISTIC_OPS} from "../statistic/statistic.service";
import {KindOfMap} from "../settings/settings.service";

export function createDefaultScenario(map: CodeMap, margin: number): Scenario {

    return {
        name: "rloc/mcc/mcc(20,40)",
        settings: {
            map: map,
            neutralColorRange: {
                from: 20,
                to: 40,
                flipped: false
            },
            areaMetric: "rloc",
            heightMetric: "mcc",
            colorMetric: "mcc",
            mode: KindOfMap.Single,
            amountOfTopLabels: 1,
            scaling: {
                x: 1, y: 1, z: 1
            },
            camera: {
                x: 0, y: 300, z: 1000
            },
            margin: margin,
            operation: STATISTIC_OPS.NOTHING,
            deltaColorFlipped: false,
            enableEdgeArrows: true,
            hideFlatBuildings: true,
            maximizeDetailPanel: false,
            invertHeight: false,
            dynamicMargin: true,
            isWhiteBackground: false,
            whiteColorBuildings: true,
            blacklist:[],
            searchPattern: ""
        },
        autoFitCamera: true
    };

}