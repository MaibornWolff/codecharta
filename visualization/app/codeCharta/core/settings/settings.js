"use strict";

import "../url/url.js";
import "../data/data.js";

import {SettingsService} from "./settingsService.js";

angular.module(
    "app.codeCharta.core.settings",
    ["app.codeCharta.core.url", "app.codeCharta.core.data"]
);

angular.module("app.codeCharta.core.settings").service(
    "settingsService", SettingsService
);

