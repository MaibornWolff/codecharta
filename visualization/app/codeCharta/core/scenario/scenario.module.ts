"use strict"

import angular from "angular"

import "../settings/settings.module"
import "../data/data.module"

import { ScenarioService } from "./scenario.service"

angular.module("app.codeCharta.core.scenario", ["app.codeCharta.core.settings", "app.codeCharta.core.data"])

angular.module("app.codeCharta.core.scenario").service("scenarioService", ScenarioService)
