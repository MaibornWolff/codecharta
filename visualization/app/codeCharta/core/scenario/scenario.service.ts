"use strict";
import {Settings} from "../settings/settings.service";
import {createDefaultScenario} from "./scenario.data";
import {ThreeOrbitControlsService} from "../../ui/codeMap/threeViewer/threeOrbitControlsService";

export interface Scenario {
    name: string;
    settings: Settings;
    autoFitCamera: boolean;
}

/**
 * Applies and manages scenarios.
 */
export class ScenarioService {

    private scenarios: Scenario[];

    /* ngInject */
    constructor(private settingsService, private dataService, private threeOrbitControlsService:ThreeOrbitControlsService) {
        this.scenarios = require("./scenarios.json");
    }

    /**
     * Applies a given scenario to the current codecharta session.
     * @param {Scenario} scenario
     */
    public applyScenario(scenario: Scenario) {
        this.settingsService.applySettings(scenario.settings);
        if(scenario.autoFitCamera){
            this.threeOrbitControlsService.autoFitTo();
        }
    }

    /**
     * Returns an array of all scenarios.
     * @returns {Scenario[]} all scenarios
     */
    public getScenarios(): Scenario[] {
        return this.scenarios.filter(s => this.isScenarioPossible(s, this.dataService._data.metrics));
    }

    public isScenarioPossible(scenario: Scenario, metrics: string[]) {
        if(!scenario || !metrics) {
            return false;
        }
        return (metrics.filter(x => x === scenario.settings.areaMetric).length > 0 &&
        metrics.filter(x => x === scenario.settings.heightMetric).length > 0 &&
        metrics.filter(x => x === scenario.settings.colorMetric).length > 0);
    }

    /**
     * Returns the default scenario.
     * @returns {Scenario} the scenario
     */
    public getDefaultScenario(): Scenario {
        return createDefaultScenario(this.settingsService.settings.map);
    }

}