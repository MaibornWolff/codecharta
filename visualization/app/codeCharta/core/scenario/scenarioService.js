"use strict";

import {Scenario} from "./model/scenario.js";
import {Settings} from "../settings/model/settings.js";
import {Scale} from "../settings/model/scale.js";
import {Range} from "../settings/model/range.js";

/**
 * Applies and manages scenarios.
 */ 
class ScenarioService {

    /* ngInject */

    /**
     * @constructor 
     * @param {SettingsService} settingsService
     * @param {DataService} dataService
     */
    constructor(settingsService, dataService) {

        /**
         * @type {SettingsService}
         */
        this.settingsService = settingsService;

        /**
         * @type {DataService}
         */
        this.dataService = dataService;

    }

    /**
     * Applies a given scenario to the current codecharta session.
     * @param {Scenario} scenario
     */
    applyScenario(scenario) {
        this.settingsService.applySettings(scenario.settings);
    }

    /**
     * Returns an array of all scenarios.
     * @returns {Scenario[]} all scenarios
     */
    getScenarios() {
        return [this.getDefaultScenario()];
    }

    /**
     * Returns the default scenario.
     * @returns {Scenario} the scenario
     */
    getDefaultScenario() {
        const defaultRange = new Range(20,40,false);
        const defaultSettings = new Settings(this.settingsService.settings.map, defaultRange, "rloc", "mcc", "mcc", false, false,1, new Scale(1,1,1), 1);
        return new Scenario("rloc/mcc/mcc(20,40)", defaultSettings);
    }

}

export {ScenarioService};