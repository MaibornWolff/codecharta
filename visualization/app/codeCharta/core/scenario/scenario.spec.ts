import {NGMock} from "../../../../mocks/ng.mockhelper";

import "./scenario.module";
import {Scenario, ScenarioService} from "./scenario.service";
import {KindOfMap, SettingsService} from "../settings/settings.service";
import {createDefaultScenario} from "./scenario.data";
import {DataService} from "../data/data.service";

/**
 * @test {ScenatioService}
 */
describe("app.codeCharta.core.scenarioService", function () {

    let scenarioService: ScenarioService,
        $scope,
        scenario: Scenario,
        defaultScenario: Scenario,
        dataService: DataService,
        settingsService: SettingsService;

    //noinspection TypeScriptUnresolvedVariable
    beforeEach(NGMock.mock.module("app.codeCharta.core.scenario"));

    //noinspection TypeScriptUnresolvedVariable
    beforeEach(NGMock.mock.inject((_scenarioService_, _settingsService_, _$rootScope_, _dataService_) => {
        scenarioService = _scenarioService_;
        settingsService = _settingsService_;
        dataService = _dataService_;
        $scope = _$rootScope_;
        scenario = {name: "testScenario", settings: settingsService.settings};
        defaultScenario = createDefaultScenario(settingsService.settings.map);
    }));

    it("should apply the settings from a given scenario", () => {
        scenarioService.applyScenario(scenario);
        expect(settingsService.settings).toBe(scenario.settings);
    });

    it("should apply the settings from a given scenario once when applyScenarioOnce is called", () => {
        scenarioService.applyScenario = jest.fn();
        scenarioService.applyScenarioOnce(scenario);
        scenarioService.applyScenarioOnce(scenario);
        expect(scenarioService.applyScenario).toHaveBeenCalledTimes(1);
    });

    it("default scenario should be rloc/mcc/mcc", () => {
        let scenario = scenarioService.getDefaultScenario();
        expect(scenario.settings.areaMetric).toBe(defaultScenario.settings.areaMetric);
        expect(scenario.settings.heightMetric).toBe(defaultScenario.settings.heightMetric);
        expect(scenario.settings.colorMetric).toBe(defaultScenario.settings.colorMetric);
    });

    it("scenarios should be scenarios from json file when all metrics are set", () => {
        dataService.data.metrics = ["rloc", "mcc", "unary", "Average Complexity*", "line_coverage", "abs_code_churn", "weeks_with_commits"];
        expect(scenarioService.getScenarios()).toEqual(require("./scenarios.json"));
    });

    it("scenarios should be filtered when not all metrics are set", () => {
        dataService.data.metrics = ["unary", "Average Complexity*"];
        expect(scenarioService.getScenarios()).toMatchSnapshot();
    });

    describe("isScenarioPossible", ()=>{

        it("should be possible", () => {
            let metrics = ["unary", "stuff"];
            let scenario = {
                settings: {
                    areaMetric: "unary",
                    heightMetric: "stuff",
                    colorMetric: "unary"
                }
            };
            expect(scenarioService.isScenarioPossible(scenario, metrics)).toBe(true);
        });

        it("should be impossible", () => {
            let metrics = ["unary", "stuff"];
            let scenario = {
                settings: {
                    areaMetric: "unary",
                    heightMetric: "stuff",
                    colorMetric: "mcc"
                }
            };
            expect(scenarioService.isScenarioPossible(scenario, metrics)).toBe(false);
        });

        it("should be impossible when params are null", () => {
            expect(scenarioService.isScenarioPossible(null, null)).toBe(false);
        });

    });

});