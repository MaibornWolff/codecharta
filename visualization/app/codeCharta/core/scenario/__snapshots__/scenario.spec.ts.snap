// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.codeCharta.core.scenarioService scenarios should be filtered when not all metrics are set 1`] = `
Array [
  Object {
    "autoFitCamera": true,
    "name": "Average Complexity*",
    "settings": Object {
      "amountOfTopLabels": 1,
      "areaMetric": "unary",
      "blacklist": Array [],
      "camera": Object {
        "x": 0,
        "y": 300,
        "z": 1000,
      },
      "colorMetric": "Average Complexity*",
      "deltaColorFlipped": false,
      "dynamicMargin": true,
      "enableEdgeArrows": true,
      "heightMetric": "Average Complexity*",
      "hideFlatBuildings": true,
      "invertHeight": false,
      "isWhiteBackground": false,
      "margin": 1,
      "maximizeDetailPanel": false,
      "mode": "KindOfMap.Single",
      "neutralColorRange": Object {
        "flipped": false,
        "from": 20,
        "to": 40,
      },
      "scaling": Object {
        "x": 1,
        "y": 1,
        "z": 1,
      },
    },
  },
]
`;
